File: 1013 - Partition Array Into Three Parts With Equal Sum.cpp
class Solution {
public:
    bool canThreePartsEqualSum(vector<int>& arr) {
        int s = 0;
        for (int a : arr) s += a;
        if (s % 3 != 0) return false;
        int ps = s / 3, i = 0;
        bool p1 = false, p2 = false;
        s = 0;
        for (; i < arr.size(); i++) {
            s += arr[i];
            if (s == ps) {
                s = 0;
                if (!p1) p1 = true;
                else if (!p2) {
                    p2 = true;
                    break;
                }
            }
        }
        return p1 && p2 && i != arr.size() - 1;
    }
};

File: 1013 - Partition Array Into Three Parts With Equal Sum.java
class Solution {
    public boolean canThreePartsEqualSum(int[] arr) {
        int sum = 0;
        for (int a : arr) {
            sum += a;
        }
        if (sum % 3 != 0) return false;
        int ps = sum / 3;
        boolean p1 = false, p2 = false;
        sum = 0;
        int i = 0;
        for (; i < arr.length; i++) {
            sum += arr[i];
            if (sum == ps) {
                sum = 0;
                if (!p1) p1 = true;
                else if (!p2) {
                    p2 = true;
                    break;
                }
            }
        }
        return p1 && p2 && i != arr.length - 1;
    }
}